src/ocrd_utils/os.py

```diff
@@ -123,19 +127,38 @@ def list_resource_candidates(executable, fname, cwd=getcwd(), moduled=None, xdg_
         candidates.append(join(moduled, fname))
     return candidates
 
-def list_all_resources(executable, moduled=None, xdg_data_home=None):
+def list_all_resources(executable : str, ocrd_tool : Optional[Dict[str, Any]] = None, moduled : Optional[str] = None, xdg_data_home : Optional[str] = None) -> List[Tuple[str,str]]:
```

Comment:
I avoid whitespaces before semicolons when `typing` to fit more parameters on a single line for the longer method signatures. Also, with `Optional[]` typing, passing `None` as a default is unnecessary; it is `None` by default.

Reply:
Ok, I saw different styles so I was confused. I also like fewer whitespace, but in the past some linters gave me a hard time because of that...

The bit about `None` default I don't understand, though. IIUC a `=None` tells Python that it's a kwarg, rather than a positional one. That should be independent of type hints. And if I use `=None`, then it *must* be `Optional` or `|None` otherwise the type guard will complain.

Reply:
> Ok, I saw different styles so I was confused. I also like fewer whitespace, but in the past some linters gave me a hard time because of that...

Weird, because internal Python libraries, when typing, use less whitespace.

> IIUC a =None tells Python that it's a kwarg, rather than a positional one. That should be independent of type hints. And if I use =None, then it must be Optional or |None otherwise the type guard will complain.

Right. `Optional[type]` was a replacement for `Union[type, None]`, but not defining the default value. My bad.